marker_api/src/context.rs

```diff
@@ -83,8 +85,10 @@ where
 /// This context will be passed to each [`LintPass`](`super::LintPass`) call to enable the user
 /// to emit lints and to retrieve nodes by the given ids.
 #[repr(C)]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
 pub struct MarkerContext<'ast> {
-    driver: &'ast DriverCallbacks<'ast>,
+    ast: AstMap<'ast>,
+    callbacks: MarkerContextCallbacks<'ast>,
```

Comment:
So fundamentlly, this changes the call site like this
```diff
- ctx.item(id)
+ ctx.ast().item(id)
```
Not sure how I feel about this. It's just the syntax organizational change. Conceptually, `MarkerContext` stores everything that one needs to query the AST/semantic information, and this separation of `AstMap` adds more boilerplate syntax to access that info.

Although It's hard to suggest better organization here. All I think of is sorting method by AST, semantics, name resolution

Reply:
Yeah, AST and semantics are the two things I would probably extract. And, most likely, one that is only for internal callbacks.

One reason for this change is to limit the number of functions on a single struct. Rustc's [TyCtxt](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#) is a worst-case example of how a context struct can get out of control. It works for rustc and has its reasons for being as it is, but I do find it difficult to navigate occasionally.

The other benefit I see is the option to group documentation. For example, the `AstMap` often returns `Option<Node>`. Here we could just document the reason once and then always just add a reference to the documentation of the main struct instead of having it on every function.